norns/clocks.md

```diff
@@ -0,0 +1,237 @@
+
+## commands
+
+- `id = clock.run( func )` &mdash; start a new coroutine with function _func_. returns _id_
+- `clock.cancel( id )` &mdash; cancel coroutine _id_ 
+- `clock.sleep( time )` &mdash; resume in _time_ seconds
+- `clock.sync( beats )` &mdash; resume at _beats_ count according to global tempo
+- `beats = clock.get_time_beats()` &mdash; returns current time in _beats_
+- `tempo = clock.get_tempo()` &mdash; returns current _tempo_
+- `beat_sec = clock.get_beat_sec()` &mdash; returns length of a single beat at current tempo in seconds
+
+## callbacks
+
+these functions can be script defined and are called by system events:
+
+- `clock.transport.start` &mdash; called by transport start
+- `clock.transport.stop` &mdash; called by transport stop
+
+## use
+
+### simple delay
+
+`clock.run` starts a new coroutine with the function you provide:
+
+```
+function later()
+  clock.sleep(2)
+  print("now awake")
+end
+
+function init()
+  print("starting now")
+  clock.run(later)
+  print("done with init")
+end
+```
+
+when executed, you will see _starting now_, _done with init_, (two seconds delay and then): _now awake_. `clock.run` starts a coroutine with the function `later`. this function immediately sleeps for two seconds, which means the `init` function resumes.
+   
+### repetition and sync
+
+you can sleep according to units of tempo (beats) by using `clock.sync` instead of `clock.sleep`.
+
+```
+engine.name = 'PolyPerc'
+
+function strum()
+  for i=1,8 do
+    engine.hz(i*100)
+    clock.sync(1/4)
+  end
+end
+
+function key(n,z)
+  if n==3 and z==1 then
+    clock.run(strum)
+  end
+end
+```
+
+this script starts a new coroutine when K3 is pressed. the `strum` function plays eight notes, sleeping for 1/4 of a beat between notes.
+
+you can change the global tempo using parameters via command or menu. see the later section.
+
+### arguments
+
+pass arguments to a clock function by appending them within the `clock.run` call:
+
+```
+engine.name = 'PolyPerc'
+
+function strum(n,speed)
+  for i=1,n do
+    clock.sync(1/speed)
+    engine.hz(i*100)
+  end
+end
+
+function key(n,z)
+  if n==3 and z==1 then
+    clock.run(strum, math.random(16), math.random(8))
+  end
+end
+```
+
+now each `strum` is executed with different arguments. note that you can overlap multiple runs by pressing the key rapidly! a new coroutine is created with each run, so you can have numerous clock processes running at once.
+
+`clock.sync` sleeps until the next subdivision specified arrives, so the timing is effectively quantized to the master tempo (which is the goal of this use case).
+
+### loop
+
+so far all of our clock functions end by themselves, by reaching the end of their own execution. we can create forever-running loops using a `while true` inner loop. these will continue running until you cancel them manually with `clock.cancel`, or by launch a new script.
```

Comment:
```suggestion
so far all of our coroutines end by themselves, by reaching the end of their own execution. we can create forever-running loops using a `while true` inner loop. these will continue running until you launch a new script or cancel them manually with `clock.cancel`.

to cancel a coroutine, you need to store its `id`. in the following script, we establish a global variable `clock_id` to capture the identifier for our running function, so that we can directly cancel it later.
```